src/passes/Heap2Local.cpp

```diff
@@ -804,11 +512,340 @@ struct Heap2LocalOptimizer {
   }
 };
 
-struct Heap2Local : public WalkerPass<PostWalker<Heap2Local>> {
+// An optimizer that handles the rewriting to turn a struct allocation into
+// locals. We run this after proving that allocation does not escape.
+//
+// TODO: Doing a single rewrite walk at the end (for all structs) would be more
+//       efficient, but it would need to be more complex.
+struct Struct2Local : PostWalker<Struct2Local> {
+  StructNew* allocation;
+  const EscapeAnalyzer& analyzer;
+  Function* func;
+  Module& wasm;
+  Builder builder;
+  const FieldList& fields;
+
+  Struct2Local(StructNew* allocation,
+               const EscapeAnalyzer& analyzer,
+               Function* func,
+               Module& wasm)
+    : allocation(allocation), analyzer(analyzer), func(func), wasm(wasm),
+      builder(wasm), fields(allocation->type.getHeapType().getStruct().fields) {
+
+    // Allocate locals to store the allocation's fields in.
+    for (auto field : fields) {
+      localIndexes.push_back(builder.addVar(func, field.type));
+    }
+
+    // Replace the things we need to using the visit* methods.
+    walk(func->body);
+
+    if (refinalize) {
+      ReFinalize().walkFunctionInModule(func, &wasm);
+    }
+  }
+
+  // Maps indexes in the struct to the local index that will replace them.
+  std::vector<Index> localIndexes;
+
+  // In rare cases we may need to refinalize, see below.
+  bool refinalize = false;
```

Comment:
Actually I intentionally put them below. We often have that pattern that arriving stuff is on top and internals appear later, like here. In particular this way shows a nice separation between the two groups. I suppose a comment could also achieve the same goal, but again, I think it might be less consistent with other code?

Reply:
I can't say I've been thinking about whether the fields are used from the constructor or not... :smile:  Maybe it is more readable that way, I'm not sure. I also don't feel very strongly here.

I already have a followup PR here, so maybe let me land this and we can consider the ordering here in that PR? This question will appear a second time there (I am adding another class much like this one).

Reply:
Yep, sgtm.